src/analyzer/protocol/tcp/TCP_Endpoint.cc

```diff
@@ -249,6 +269,48 @@ int TCP_Endpoint::DataSent(double t, uint64 seq, int len, int caplen,
 	return status;
 	}
 
+void TCP_Endpoint::EnqueuePacket(int len, const u_char* data, bool is_orig,
+				 uint64 seq, const IP_Hdr* ip, int caplen, uint32 ack_seq)
+	{
+	auto tcp_p = new TCP_Packet(len, data, is_orig, seq, ip, caplen, ack_seq);
+
+	// ensure tcp-packet was created properly.
+	if (tcp_p and (tcp_p->ip and tcp_p->data))
+		tcp_pq.emplace(tcp_p);	  // enque pkt for deferred replay.
```

Comment:
This logic was suggesting a possible leak -- if that conditional is false, nothing takes ownership to free `tcp_p`, but reading more closely, that conditional is just always true and can be removed.  There were a few other spots that check for non-null pointer after doing `new` (or also before `delete`) that aren't necessary.

Reply:
yeah, tcp-packet could be allocated but allocation for ip or data could fail. In those cases, tcp packet is not valid and should be discarded. Anyways, I've changed the implementation to use shared_ptr and unique_ptr. I'll send a new PR early next week.